README.md
<img align="left" width="120" height="120" src="https://github.com/redsolver/skydroid/raw/master/assets/icon/icon.png" alt="App Icon">

# SkyDroid • Decentralized App Store

[![Chat on Matrix](https://img.shields.io/matrix/skydroid:matrix.org?style=for-the-badge)](https://matrix.to/#/#skydroid:matrix.org)
![GPL-3.0 License](https://img.shields.io/github/license/redsolver/skydroid?style=for-the-badge)

[![Help Translate](https://weblate.bubu1.eu/widgets/skydroid/-/skydroid-app/287x66-black.png)](https://weblate.bubu1.eu/engage/skydroid/)


SkyDroid is a decentralized domain-based **Android** App Store which offers easy and fast app distribution, discoverability and security through collections, multi-language support, multiple Themes, Search and Filter options, nice error handling and of course a good user experience.

Related repos are the [F-Droid bridge](https://github.com/redsolver/skydroid-fdroid-bridge) and the [Multi-DoH Server](https://github.com/redsolver/multi-doh-server).

SkyDroid uses native code from https://github.com/Aefyr/SAI (GPL-3.0).

## Screenshots

<p>
  <img src="https://github.com/redsolver/skydroid/raw/master/screenshots/screen1.jpg" width="250">
  <img src="https://github.com/redsolver/skydroid/raw/master/screenshots/screen2.jpg" width="250">
  <img src="https://github.com/redsolver/skydroid/raw/master/screenshots/screen3.jpg" width="250">
</p>

## Install + First Steps

1. Visit https://get.skydroid.app on your Android device
2. Open the downloaded APK file and install it
3. Open SkyDroid and go to "Collections"
4. Add some recommended Collections
5. Navigate back to "Apps" and enjoy!

## How to publish your own app

See https://skydroid.dev

## Video Demo

Here's a little demo of the app and how it works: https://youtu.be/MTSrz3Jb778?t=2461

## Why it's important

Fair and secure app distribution is currently more important than ever. (see the current news about the Apple App Store and Google Play)

SkyDroid aims to fix this problem on Android by making direct app distribution decentralized, convenient to use, affordable and secure.

## Technical Details

### Apps

Every app on SkyDroid is offered through a specific domain. [Handshake Domains](https://handshake.org/) are also fully supported to allow full decentralization.

For example the domain `noteless.redsolver` has a `TXT` record with the following content:

`skydroid-app=1+AADbpx41U1UCRcIhSHvzRAgt8LJYaDlxiLyqHnPj8ckXAA+0a2e07bb2666409ceb6f49072e296d6ca4f2050af098da1cf6d17fd09b49e6cc`

The `skydroid-app=` part tells SkyDroid, that this Domain/Name contains an App.

The `1` is the version of this record format.

The `AADbpx41U1UCRcIhSHvzRAgt8LJYaDlxiLyqHnPj8ckXAA` part is a Skylink (explained in a moment) which points to the metadata file of the app.

The `0a2e07bb2666409ceb6f49072e296d6ca4f2050af098da1cf6d17fd09b49e6cc` part is a sha256 hash of the metadata file to ensure the integrity.

Ok, but what is a Skylink?

A Skylink points to a file on the Sia Skynet. The Sia Skynet is a decentralized CDN and file sharing platform for devs.

It works like this: There are multiple so-called "portals" to the Skynet. (for example https://siasky.net/ or https://www.siacdn.com/, anyone can host one!)

A file which is uploaded to one portal, can be downloaded from any other portal!

This enables fully decentralized file-sharing, because the uploader can use any portal or even directly upload to Skynet and also the user can choose between any of the portals!

Skylinks **must** be used for the metadata files and **can** be used for the app icon, images and the APK files.

Ok, so back to the metadata file!

If you add `noteless.redsolver` in SkyDroid, the apps checks the TXT records of the name like explained above and then downloads the metadata file.

You can choose any Skynet portal of your choice in the SkyDroid settings.

The metadata file is checked against the hash and then the app is shown in the UI.

When a SkyDroid user wants to check for updates of all apps, only the `TXT` records need to be checked.
Only if a new metadata hash is found, the metadata file is downloaded again and applied.

### Collections

A collection in SkyDroid has two main goals:

- Offer a way to discover apps
- Offer a decentralized trust-system for apps

A collection metadata file is loaded exactly like an app metadata file from a domain name (for example `redsolver`),
but it uses `skydroid-collection=` instead in the `TXT` record.

#### Discover Apps

A collection can be made by anyone.

It usually contain's multiple curated Domains/Names from SkyDroid apps which the collection author wants to recommend.

For example, a developer could have a collection of his own apps and some other apps he found and finds nice.

When the SkyDroid user adds the collection, every app in it is added as well.

Some collections are "recommended" by me in the app to get the user started fast, but any collection can be added easily by simply pressing the "Add" button and entering the Domain/Name.

#### Decentralized Trust-System

Collections can also "verify" apps they list. That basically means putting the hash of the app metadata to the name.

This allows security researchers or companies to check the app and mark it as "secure".

This system can also be used for other "verification" to for example check if an app is good or not.

The user can see how many and which collections verified the app in the current state on the "App Page".

### Metadata file examples

The `redsolver` collection metadata file (can be JSON or YAML):

```yaml
title: red's collection
description: This collection contains every app available via SkyDroid that I'm aware of. Some apps are verified.
icon: sia://PAGUwiKmHy_83Att8NssAMj79PF1V8g5x_B2lKyThFhKig

apps:
  - name: noteless.redsolver
    verifiedMetadataHashes:
      [
        f7922a73001a8838db29aa4eea5bc91244c07b65fd9a2f1bad16b6a491308af4,
        4bd9b5ad567784b4defa02a892794b28b4acc0b8e32b6c2e996f955280cfed02,
        f06c6781e7964ab458903594189734216a06f71bc820d2816c3f5347cff72fd5,
        0a2e07bb2666409ceb6f49072e296d6ca4f2050af098da1cf6d17fd09b49e6cc,
      ]
  - name: skydroid.app
    verifiedMetadataHashes:
      [bdcbc30b582a078deaea5c6c12f8d265ca22922e5143bb63974e91871de343fe]

```

The `noteless.redsolver` app metadata file (can be JSON or YAML):

```yaml
categories:
  - Writing
license: MIT
authorName: redsolver
authorEmail: info@redsolver.net
sourceCode: https://github.com/redsolver/noteless
issueTracker: https://github.com/redsolver/noteless/issues
changelog: https://github.com/redsolver/noteless/blob/HEAD/CHANGELOG.md

name: Noteless
packageName: net.redsolver.noteless

icon: https://github.com/redsolver/noteless/raw/master/assets/icon/icon.png

localized:
  en-US:
    description: |-
      Features

      * Markdown-optimized editor with syntax highlighting
      * Supports Github Flavored Markdown, KaTeX and mermaidjs for diagrams
      * Tags for organizing your notes
      * Pin, Star and sort your notes by title or different dates
      * Very themable - dark/light mode and accent color
      * Full-text search
      * File Attachments that can be embedded into a note
      * Multi-Note Editing
      * Slide actions for easier editing
      * Tutorial notes which explain how to use the app
    summary: A markdown note-taking app for mobile devices.
    whatsNew: "First release on SkyDroid, yay!"
    phoneScreenshotsBaseUrl: https://github.com/redsolver/noteless/raw/master/screenshots/
    phoneScreenshots:
      - screen1.png
      - screen2.png
      - screen3.png
      - screen4.png
      - screen5.png
      - screen6.png
      - screen7.png
      - screen8.png
      - screen9.png
      - screen10.png
      - screen11.png
      - screen12.png

builds:
  - versionName: 0.3.2
    versionCode: 32
    sha256: 0a5d15f00554e982e37ed9bdda41d17304e8f056913a8d0921a11d81844578e9
    apkLink: sia://AADTpnQ7qv4aFsJYTJECT8e8428kGgLmwLUWgqdycvquFg
  - versionName: 1.0.0
    versionCode: 100
    sha256: d56dccb20c45e84089560a8d7611eeb9a55a15647fa06aee822364f39c753a69
    apkLink: sia://AABBdXkn8Ogia55qK2vvYLRecYY2U-sKyTx0SoD8c6DOKQ

currentVersionName: 1.0.0
currentVersionCode: 100

added: 1596499200000
lastUpdated: 1596737665000
```

The screenshots and app icon can also be hosted on the Sia Skynet. (Just use `sia://` instead of `https://`, it gets automatically replaced with the selected Skynet Portal)

## How to publish your own app (manual method)

- Copy the contents of `minimal-app-template.yaml` to a new YAML file on your computer
- Edit all values to match your app
  - You can upload your APK file and media files to https://siasky.net or just use a normal https link if they are already hosted on GitHub for example
- Get the hash of the metadata file by running `sha256sum your-app-file.yaml` in your terminal or using a sha256 website like [this one](https://emn178.github.io/online-tools/sha256.html)
- Upload your YAML file to https://siasky.net

Build a TXT Record like this:

`skydroid-app=1+YOUR_SKYLINK_FROM_SKYNET+YOUR_SHA256_HASH`

Example: `skydroid-app=1+AADbpx41U1UCRcIhSHvzRAgt8LJYaDlxiLyqHnPj8ckXAA+0a2e07bb2666409ceb6f49072e296d6ca4f2050af098da1cf6d17fd09b49e6cc`

- Add this TXT Record to a domain/name of your choice. (You can use a normal or a Handshake domain)
- Add you domain/name in the SkyDroid app with the "Plus"-button in the bottom right

## About

This app was initially developed as a submission to the [‘Own The Internet’ Hackathon](https://gitcoin.co/hackathon/own-the-internet).

## Planned features

- QR-Code Sharing of Apps
- Some sort of decentralized Rating System (maybe via Mastodon)
- Maybe use [Shizuku service](https://github.com/RikkaApps/Shizuku) like https://github.com/Aefyr/SAI to enable automatic background updates (if verified).
 
Oh, and the app is written in Flutter - making the SkyDroid App Store available on Windows, macOS or Linux is possible.


CHANGELOG.md
# Changelog

## 0.5.5

- Added support for different versionCodes for different ABIs for the same versionName (This fixes the bug that some apps can't be installed)

## 0.5.4

- Added new languages and updated translations (Thanks to all contributors!)
- Added new URI scheme to make app links more reliable
- Small improvements and bugfixes

## 0.5.3

Important bug fix: Wrong language used instead of English on unsupported locales

## 0.5.2

- Updated translations
- Changed app icon
- Updated recommended collections
- Fixed some small bugs

## 0.5.1

- Skydroid now prefers the `_skydroid` subdomain for `TXT` records to prevent collision with existing records (Both ways are still supported)
- Changed domain of the Shizuku Service App to bridged IzzyOnDroid repo (https://collection.skydroid.app/izzyondroid)

## 0.4.1

- Added batch selection and processing actions
- Added deeplink support for collections
- Added automatic cache cleanup
- Improved error messages
- Translated in Czech

## 0.3.0

- Translated in English, German, French and Dutch
- Shizuku Service Support for easier app installation
- Split-APK support for up to 3x smaller download sizes
- Changed accent color
- Fixed some bugs and added some optimizations

## 0.2.4

Fixed error when categories are missing in the metadata

## 0.2.3

Performance optimizations to limit the number of concurrent network requests

## 0.2.2

Fixed Out of Memory when downloading APKs bug on devices with less RAM

## 0.2.0

You can now share apps via the to.skydroid.app domain and open them directly in SkyDroid!

## 0.1.4

- Changed Domain of F-Droid Collection to "fdroid-app"
- Fixed a small bug which marked an app downgrade as an update in some cases

## 0.1.3

- Added option to remove all apps from a specific collection

## 0.1.2

- Fixed some bugs


lib/model/app.g.dart
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'app.dart';

// **************************************************************************
// TypeAdapterGenerator
// **************************************************************************

class AppAdapter extends TypeAdapter<App> {
  @override
  final int typeId = 1;

  @override
  App read(BinaryReader reader) {
    final numOfFields = reader.readByte();
    final fields = <int, dynamic>{
      for (int i = 0; i < numOfFields; i++) reader.readByte(): reader.read(),
    };
    return App()
      ..categories = (fields[1] as List)?.cast<String>()
      ..license = fields[2] as String
      ..authorName = fields[3] as String
      ..authorEmail = fields[4] as String
      ..sourceCode = fields[5] as String
      ..issueTracker = fields[6] as String
      ..changelog = fields[7] as String
      ..name = fields[8] as String
      ..packageName = fields[9] as String
      ..localized = (fields[20] as Map)?.map((dynamic k, dynamic v) =>
          MapEntry(k as String, (v as Map)?.cast<dynamic, dynamic>()))
      ..summary = fields[30] as String
      ..description = fields[31] as String
      ..whatsNew = fields[32] as String
      ..icon = fields[22] as String
      ..builds = (fields[11] as List)?.cast<Build>()
      ..currentVersionName = fields[12] as String
      ..currentVersionCode = fields[13] as int
      ..added = fields[14] as int
      ..lastUpdated = fields[15] as int
      ..webSite = fields[16] as String
      ..translation = fields[17] as String
      ..antiFeatures = (fields[24] as List)?.cast<String>()
      ..donate = fields[18] as String
      ..bitcoin = fields[19] as String
      ..litecoin = fields[25] as String
      ..liberapay = fields[26] as String
      ..openCollective = fields[28] as String
      ..requirements = fields[29] as String
      ..metadataSrcHash = fields[250] as String;
  }

  @override
  void write(BinaryWriter writer, App obj) {
    writer
      ..writeByte(29)
      ..writeByte(1)
      ..write(obj.categories)
      ..writeByte(2)
      ..write(obj.license)
      ..writeByte(3)
      ..write(obj.authorName)
      ..writeByte(4)
      ..write(obj.authorEmail)
      ..writeByte(5)
      ..write(obj.sourceCode)
      ..writeByte(6)
      ..write(obj.issueTracker)
      ..writeByte(7)
      ..write(obj.changelog)
      ..writeByte(8)
      ..write(obj.name)
      ..writeByte(9)
      ..write(obj.packageName)
      ..writeByte(20)
      ..write(obj.localized)
      ..writeByte(30)
      ..write(obj.summary)
      ..writeByte(31)
      ..write(obj.description)
      ..writeByte(32)
      ..write(obj.whatsNew)
      ..writeByte(22)
      ..write(obj.icon)
      ..writeByte(11)
      ..write(obj.builds)
      ..writeByte(12)
      ..write(obj.currentVersionName)
      ..writeByte(13)
      ..write(obj.currentVersionCode)
      ..writeByte(14)
      ..write(obj.added)
      ..writeByte(15)
      ..write(obj.lastUpdated)
      ..writeByte(16)
      ..write(obj.webSite)
      ..writeByte(17)
      ..write(obj.translation)
      ..writeByte(24)
      ..write(obj.antiFeatures)
      ..writeByte(18)
      ..write(obj.donate)
      ..writeByte(19)
      ..write(obj.bitcoin)
      ..writeByte(25)
      ..write(obj.litecoin)
      ..writeByte(26)
      ..write(obj.liberapay)
      ..writeByte(28)
      ..write(obj.openCollective)
      ..writeByte(29)
      ..write(obj.requirements)
      ..writeByte(250)
      ..write(obj.metadataSrcHash);
  }

  @override
  int get hashCode => typeId.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is AppAdapter &&
          runtimeType == other.runtimeType &&
          typeId == other.typeId;
}

class BuildAdapter extends TypeAdapter<Build> {
  @override
  final int typeId = 3;

  @override
  Build read(BinaryReader reader) {
    final numOfFields = reader.readByte();
    final fields = <int, dynamic>{
      for (int i = 0; i < numOfFields; i++) reader.readByte(): reader.read(),
    };
    return Build()
      ..versionName = fields[1] as String
      ..versionCode = fields[2] as int
      ..sha256 = fields[3] as String
      ..apkLink = fields[4] as String
      ..abis = (fields[11] as Map)?.cast<String, ABISpecificBuild>()
      ..size = fields[5] as int
      ..minSdkVersion = fields[7] as int
      ..targetSdkVersion = fields[8] as int
      ..added = fields[9] as int
      ..permissions = (fields[10] as List)?.cast<String>();
  }

  @override
  void write(BinaryWriter writer, Build obj) {
    writer
      ..writeByte(10)
      ..writeByte(1)
      ..write(obj.versionName)
      ..writeByte(2)
      ..write(obj.versionCode)
      ..writeByte(3)
      ..write(obj.sha256)
      ..writeByte(4)
      ..write(obj.apkLink)
      ..writeByte(11)
      ..write(obj.abis)
      ..writeByte(5)
      ..write(obj.size)
      ..writeByte(7)
      ..write(obj.minSdkVersion)
      ..writeByte(8)
      ..write(obj.targetSdkVersion)
      ..writeByte(9)
      ..write(obj.added)
      ..writeByte(10)
      ..write(obj.permissions);
  }

  @override
  int get hashCode => typeId.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is BuildAdapter &&
          runtimeType == other.runtimeType &&
          typeId == other.typeId;
}

class ABISpecificBuildAdapter extends TypeAdapter<ABISpecificBuild> {
  @override
  final int typeId = 6;

  @override
  ABISpecificBuild read(BinaryReader reader) {
    final numOfFields = reader.readByte();
    final fields = <int, dynamic>{
      for (int i = 0; i < numOfFields; i++) reader.readByte(): reader.read(),
    };
    return ABISpecificBuild()
      ..apkLink = fields[1] as String
      ..sha256 = fields[2] as String
      ..versionName = fields[3] as String
      ..versionCode = fields[4] as int;
  }

  @override
  void write(BinaryWriter writer, ABISpecificBuild obj) {
    writer
      ..writeByte(4)
      ..writeByte(1)
      ..write(obj.apkLink)
      ..writeByte(2)
      ..write(obj.sha256)
      ..writeByte(3)
      ..write(obj.versionName)
      ..writeByte(4)
      ..write(obj.versionCode);
  }

  @override
  int get hashCode => typeId.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ABISpecificBuildAdapter &&
          runtimeType == other.runtimeType &&
          typeId == other.typeId;
}

// **************************************************************************
// JsonSerializableGenerator
// **************************************************************************

App _$AppFromJson(Map json) {
  return App()
    ..categories =
        (json['categories'] as List)?.map((e) => e as String)?.toList()
    ..license = json['license'] as String
    ..authorName = json['authorName'] as String
    ..authorEmail = json['authorEmail'] as String
    ..sourceCode = json['sourceCode'] as String
    ..issueTracker = json['issueTracker'] as String
    ..changelog = json['changelog'] as String
    ..name = json['name'] as String
    ..packageName = json['packageName'] as String
    ..localized = (json['localized'] as Map)?.map(
      (k, e) => MapEntry(k as String, e as Map),
    )
    ..localizedPhoneScreenshotsCache =
        (json['localizedPhoneScreenshotsCache'] as List)
            ?.map((e) => e as String)
            ?.toList()
    ..summary = json['summary'] as String
    ..description = json['description'] as String
    ..whatsNew = json['whatsNew'] as String
    ..icon = json['icon'] as String
    ..builds = (json['builds'] as List)
        ?.map((e) => e == null ? null : Build.fromJson(e as Map))
        ?.toList()
    ..currentVersionName = json['currentVersionName'] as String
    ..currentVersionCode = json['currentVersionCode'] as int
    ..added = json['added'] as int
    ..lastUpdated = json['lastUpdated'] as int
    ..webSite = json['webSite'] as String
    ..translation = json['translation'] as String
    ..antiFeatures =
        (json['antiFeatures'] as List)?.map((e) => e as String)?.toList()
    ..donate = json['donate'] as String
    ..bitcoin = json['bitcoin'] as String
    ..litecoin = json['litecoin'] as String
    ..liberapay = json['liberapay'] as String
    ..openCollective = json['openCollective'] as String
    ..requirements = json['requirements'] as String
    ..metadataSrcHash = json['metadataSrcHash'] as String;
}

Map<String, dynamic> _$AppToJson(App instance) => <String, dynamic>{
      'categories': instance.categories,
      'license': instance.license,
      'authorName': instance.authorName,
      'authorEmail': instance.authorEmail,
      'sourceCode': instance.sourceCode,
      'issueTracker': instance.issueTracker,
      'changelog': instance.changelog,
      'name': instance.name,
      'packageName': instance.packageName,
      'localized': instance.localized,
      'localizedPhoneScreenshotsCache': instance.localizedPhoneScreenshotsCache,
      'summary': instance.summary,
      'description': instance.description,
      'whatsNew': instance.whatsNew,
      'icon': instance.icon,
      'builds': instance.builds,
      'currentVersionName': instance.currentVersionName,
      'currentVersionCode': instance.currentVersionCode,
      'added': instance.added,
      'lastUpdated': instance.lastUpdated,
      'webSite': instance.webSite,
      'translation': instance.translation,
      'antiFeatures': instance.antiFeatures,
      'donate': instance.donate,
      'bitcoin': instance.bitcoin,
      'litecoin': instance.litecoin,
      'liberapay': instance.liberapay,
      'openCollective': instance.openCollective,
      'requirements': instance.requirements,
      'metadataSrcHash': instance.metadataSrcHash,
    };

Build _$BuildFromJson(Map json) {
  return Build()
    ..versionName = json['versionName'] as String
    ..versionCode = json['versionCode'] as int
    ..sha256 = json['sha256'] as String
    ..apkLink = json['apkLink'] as String
    ..abis = (json['abis'] as Map)?.map(
      (k, e) => MapEntry(
          k as String, e == null ? null : ABISpecificBuild.fromJson(e as Map)),
    )
    ..size = json['size'] as int
    ..minSdkVersion = json['minSdkVersion'] as int
    ..targetSdkVersion = json['targetSdkVersion'] as int
    ..added = json['added'] as int
    ..permissions =
        (json['permissions'] as List)?.map((e) => e as String)?.toList();
}

Map<String, dynamic> _$BuildToJson(Build instance) => <String, dynamic>{
      'versionName': instance.versionName,
      'versionCode': instance.versionCode,
      'sha256': instance.sha256,
      'apkLink': instance.apkLink,
      'abis': instance.abis,
      'size': instance.size,
      'minSdkVersion': instance.minSdkVersion,
      'targetSdkVersion': instance.targetSdkVersion,
      'added': instance.added,
      'permissions': instance.permissions,
    };

ABISpecificBuild _$ABISpecificBuildFromJson(Map json) {
  return ABISpecificBuild()
    ..apkLink = json['apkLink'] as String
    ..sha256 = json['sha256'] as String
    ..versionName = json['versionName'] as String
    ..versionCode = json['versionCode'] as int;
}

Map<String, dynamic> _$ABISpecificBuildToJson(ABISpecificBuild instance) =>
    <String, dynamic>{
      'apkLink': instance.apkLink,
      'sha256': instance.sha256,
      'versionName': instance.versionName,
      'versionCode': instance.versionCode,
    };
